### PR TITLE
Asterisk update

### PR DESCRIPTION
**Pull request by jkroonza**
There is one outstanding QA issue I'm not sure how to solve:

 * QA Notice: The following shared libraries lack NEEDED entries
 * /usr/lib64/libasteriskssl.so.1

So no merging yet.

Here is the really odd thing (file is saying *dynamic*, ldd is saying *static*):

jkroon@plastiekpoot /usr/lib64 $ ldd libasteriskssl.so.1 
        statically linked
jkroon@plastiekpoot /usr/lib64 $ file libasteriskssl.so.1
libasteriskssl.so.1: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, stripped
jkroon@plastiekpoot /usr/lib64 $ objdump -p libasteriskssl.so.1

libasteriskssl.so.1:     file format elf64-x86-64

Program Header:
    LOAD off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**12
         filesz 0x00000000000004a0 memsz 0x00000000000004a0 flags r--
    LOAD off    0x0000000000001000 vaddr 0x0000000000001000 paddr 0x0000000000001000 align 2**12
         filesz 0x000000000000015d memsz 0x000000000000015d flags r-x
    LOAD off    0x0000000000002000 vaddr 0x0000000000002000 paddr 0x0000000000002000 align 2**12
         filesz 0x00000000000000e8 memsz 0x00000000000000e8 flags r--
    LOAD off    0x0000000000002e30 vaddr 0x0000000000003e30 paddr 0x0000000000003e30 align 2**12
         filesz 0x0000000000000200 memsz 0x0000000000000208 flags rw-
 DYNAMIC off    0x0000000000002e50 vaddr 0x0000000000003e50 paddr 0x0000000000003e50 align 2**3
         filesz 0x0000000000000190 memsz 0x0000000000000190 flags rw-
EH_FRAME off    0x000000000000201c vaddr 0x000000000000201c paddr 0x000000000000201c align 2**2
         filesz 0x0000000000000034 memsz 0x0000000000000034 flags r--
   STACK off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**4
         filesz 0x0000000000000000 memsz 0x0000000000000000 flags rw-
   RELRO off    0x0000000000002e30 vaddr 0x0000000000003e30 paddr 0x0000000000003e30 align 2**0
         filesz 0x00000000000001d0 memsz 0x00000000000001d0 flags r--

Dynamic Section:
  SONAME               libasteriskssl.so.1
  INIT                 0x0000000000001000
  FINI                 0x0000000000001154
  INIT_ARRAY           0x0000000000003e30
  INIT_ARRAYSZ         0x0000000000000010
  FINI_ARRAY           0x0000000000003e40
  FINI_ARRAYSZ         0x0000000000000010
  GNU_HASH             0x0000000000000200
  STRTAB               0x00000000000002e8
  SYMTAB               0x0000000000000228
  STRSZ                0x00000000000000ac
  SYMENT               0x0000000000000018
  PLTGOT               0x0000000000004000
  PLTRELSZ             0x0000000000000030
  PLTREL               0x0000000000000007
  JMPREL               0x0000000000000470
  RELA                 0x0000000000000398
  RELASZ               0x00000000000000d8
  RELAENT              0x0000000000000018
  RELACOUNT            0x0000000000000005

### CODE REVIEW

**Comment by trofi** [comment]
> QA Notice: The following shared libraries lack NEEDED entries /usr/lib64/libasteriskssl.so.1
I think it's a false positive entry of gentoo's QA check. The library is built normally as:
```
$ x86_64-pc-linux-gnu-gcc  -o libasteriskssl.so.1 -pthread -Wl,-O1 -Wl,--as-needed -Wl,--hash-style=gnu   -Wl,-soname=libasteriskssl.so.1 -shared -Wl,--version-script,libasteriskssl.exports,--warn-common -Wl,-O1 -Wl,--as-needed -Wl,--hash-style=gnu libasteriskssl.o  -lssl -lcrypto
```

Note: `-Wl,--as-needed` removes all the library dependencies. I think it happens because that library is effectively a no-op when `defined(LIBRESSL_VERSION_NUMBER) || OPENSSL_VERSION_NUMBER < 0x10100000L`: https://github.com/asterisk/asterisk/blob/master/main/libasteriskssl.c#L34 . 

The resulting binary exports one `ast_ssl_init` symbol:
```
$ nm -D libasteriskssl.so.1
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
                 w __cxa_finalize
                 w __gmon_start__
                 U ast_register_file_version
0000000000001150 T ast_ssl_init
                 U ast_unregister_file_version
```

```c
int ast_ssl_init(void)
{
	return 0;
}
```
 and depends on nothing. Not even `libc.so.6`.

You can suppress a warning with `QA_DT_NEEDED=` variable setting in the ebuild.

**Comment by jkroonza** [comment]
> You can suppress a warning with `QA_DT_NEEDED=` variable setting in the ebuild.

Done.  Pushed.

And I can confirm the theory.

Against newer OpenSSL (where the stuff is fixed) it really operates as described.

On older OpenSSL or LibreSSL the exports kicks in, so on those hosts lddtree actually presents:

arthur ~ # equery list -ipo openssl
 * Searching for openssl ...
[I--] [??] dev-libs/openssl-1.0.2r:0
[-P-] [  ] dev-libs/openssl-1.0.2t-r1:0
[-P-] [ ~] dev-libs/openssl-1.1.0l:0/1.1
[-P-] [  ] dev-libs/openssl-1.1.1d-r2:0/1.1
arthur ~ # lddtree /usr/lib64/libasteriskssl.so.1 
libasteriskssl.so.1 => /usr/lib64/libasteriskssl.so.1 (interpreter => none)
    libcrypto.so.1.0.0 => /usr/lib64/libcrypto.so.1.0.0
        libdl.so.2 => /lib64/libdl.so.2
            ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2
        libz.so.1 => /lib64/libz.so.1
    libpthread.so.0 => /lib64/libpthread.so.0
    libc.so.6 => /lib64/libc.so.6
arthur ~ #

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64 ~ppc ~x86"
+
+IUSE_VOICEMAIL_STORAGE="
+	+voicemail_storage_file
+	voicemail_storage_odbc
+	voicemail_storage_imap
+"
+IUSE="${IUSE_VOICEMAIL_STORAGE} alsa bluetooth calendar +caps cluster curl dahdi debug doc freetds gtalk http iconv ilbc xmpp ldap libedit libressl lua mysql newt +samples odbc osplookup oss pjproject portaudio postgres radius selinux snmp span speex srtp static statsd syslog vorbis"
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: `xmpp` seems out of place in this list :)
What exaclty does `oss` do, is it still supported?

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+IUSE_EXPAND="VOICEMAIL_STORAGE"
+REQUIRED_USE="gtalk? ( xmpp )
+	^^ ( ${IUSE_VOICEMAIL_STORAGE/+/} )
+	voicemail_storage_odbc? ( odbc )
+"
+
+PATCHES=( "${WORKDIR}/asterisk-patchset" )
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: You should indicate the version this patcshet was introduced in, and preferably try to get (or as much of) it upstreamed.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+DEPEND="acct-user/asterisk
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Are these really required during build-time? There aren't many examples of these, but usually users and groups are needed when running the program.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+	acct-group/asterisk
+
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: You can remove this empty line.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	dev-db/sqlite:3
+	dev-libs/popt
+	dev-libs/jansson
+	dev-libs/libxml2
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Add a slot operator similarily to sqlite above.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+	sys-libs/ncurses:0=
+	sys-libs/zlib
+
+	!libressl? ( dev-libs/openssl:0= )
+	libressl? ( dev-libs/libressl:0= )
+
+	alsa? ( media-libs/alsa-lib )
+	bluetooth? ( net-wireless/bluez )
+	calendar? (
+		net-libs/neon
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: If the compiler links to this lib, you should add a subslot binder.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+		dev-libs/libical
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Similarily.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	menuselect/menuselect --enable cdr_sqlite3_custom menuselect.makeopts
+	menuselect/menuselect --enable cel_sqlite3_custom menuselect.makeopts
+
+	# The others are based on USE-flag settings
+	use_select() {
+		local state=$(use "$1" && echo enable || echo disable)
+		shift # remove use from parameters
+
+		while [[ -n $1 ]]; do
+			menuselect/menuselect --${state} "$1" menuselect.makeopts
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Looks like it needs `|| die`.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+	use_select speex        {codec,func}_speex
+	use_select srtp         res_srtp
+	use_select statsd       res_statsd res_{endpoint,chan}_stats
+	use_select syslog       cdr_syslog
+	use_select vorbis       format_ogg_vorbis
+
+	# Voicemail storage ...
+	for vmst in ${IUSE_VOICEMAIL_STORAGE/+/}; do
+		if use ${vmst}; then
+			menuselect/menuselect --enable $(echo ${vmst##*_} | tr '[:lower:]' '[:upper:]')_STORAGE menuselect.makeopts
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Similarily.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+		fi
+	done
+
+	if use debug; then
+		for o in DONT_OPTIMIZE DEBUG_THREADS BETTER_BACKTRACES; do
+			menuselect/menuselect --enable $o menuselect.makeopts
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: And here. You should probably add curly brackets around `o` as well.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+		insinto /etc/radiusclient/
+		doins contrib/dictionary.digium
+	fi
+	diropts -m 0750 -o root -g asterisk
+	keepdir	/etc/asterisk
+	if use samples; then
+		emake NOISY_BUILD=yes DESTDIR="${D}" samples
+		for conffile in "${ED}/etc/asterisk/"*.*
+		do
+			chown root:root $conffile
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: `fowners` and curly brackets.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+			chmod 0644 $conffile
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: `fperms` and curly brackets.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+		done
+		einfo "Sample files have been installed"
+	else
+		einfo "Skipping installation of sample files..."
+		rm -f  "${ED}"/var/lib/asterisk/mohmp3/* || die
+		rm -f  "${ED}"/var/lib/asterisk/sounds/demo-* || die
+		rm -f  "${ED}"/var/lib/asterisk/agi-bin/* || die
+		rm -f  "${ED}"/etc/asterisk/* || die
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Drop the `-f`, you should know what you're removing and when.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+	fi
+	rm -rf "${ED}"/var/spool/asterisk/voicemail/default || die
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: Same here.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+	systemd_newtmpfilesd "${FILESDIR}"/asterisk.tmpfiles.conf asterisk.conf
+	systemd_install_serviced "${FILESDIR}"/asterisk.service.conf
+
+	# install the upgrade documentation
+	#
+	dodoc UPGRADE* BUGS CREDITS
+
+	# install extra documentation
+	#
+	if use doc
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: `if use doc; then`

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+
+	# install logrotate snippet; bug #329281
+	#
+	insinto /etc/logrotate.d
+	newins "${FILESDIR}/1.6.2/asterisk.logrotate4" asterisk
+
+	# Asterisk installs a few folders that's empty by design,
+	# but still required.  This finds them, and marks them for
+	# portage.
+	for d in $(find "${ED}"/var -type d -empty); do
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: `find` needs to `|| die`.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+	use_select bluetooth    chan_mobile
+	use_select calendar     res_calendar res_calendar_{caldav,ews,exchange,icalendar}
+	use_select cluster      res_corosync
+	use_select curl         func_curl res_config_curl res_curl
+	use_select dahdi        app_dahdiras app_meetme chan_dahdi codec_dahdi res_timing_dahdi
+	use_select freetds      {cdr,cel}_tds
+	use_select gtalk        chan_motif
+	use_select http         res_http_post
+	use_select iconv        func_iconv
+	use_select xmpp         res_xmpp
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: If you want to sort this alphabetically too...

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 1 above)
Comment (jkroonza) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: It builds chan_oss, which is still alive.  I've personally never used and don't mind ditching it as "not supported".  It still builds ... which almost comes as a surprise.

xmpp seems out of place but it's not :).  It's used to integrate with gtalk which uses the xmpp protocol, as well as provides for publishing status/availability information to XMPP in general.  Not sure if we have any actual users currently.  I've been threatening to mess with it too and have another project upcoming where it may be useful.

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 2 above)
Comment (jkroonza) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: Patches are inherited.  They're mostly as old as the mountains (ie, we've been carrying them for longer than I care to admit).  Most of the patches I cooked (barring 02-dahdiras-without-root and 05-dundi-null-dereference.patch) has already been merged upstream and I'll try to get these upstream too.  That obviously won't make it for any of the current versions of asterisk.

Ok, so given that we have an existing patch set, what would normally be the mechanism to indicate when the patches were introduced, and when the specific grouping was introduced?

When putting patches in ${FILESDIR} I now name then ${PV}-shortname.patch|diff and then 

With only four patches left (although I've got about 10 more that I manage out of tree that possibly doesn't even belong upstream or even in gentoo) - any reason to not include them in-tree in the above manner?

I guess I could do ${ORDER}-${PV}-shortname.patch inside the tarrball and rename the tarball to gentoo-asterisk-patchset-${PV).tar.bz2 - would this satisfy your requirements?

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 3 above)
Comment (jkroonza) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: I believe since files (and folders) are installed as the asterisk user/group that DEPEND is correct.

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 2 above)
Comment (jkroonza) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: Oh wait, the patchset here just points to the folder created after extracting the archive, which is versioned.  Comments still relevant.

File: acct-user/asterisk/asterisk-0.ebuild [hunk 18]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+inherit acct-user
+
+DESCRIPTION="User for asterisk"
+ACCT_USER_ID=242
+ACCT_USER_HOME=/var/lib/asterisk
```
Comment (juippis) [inline acct-user/asterisk/asterisk-0.ebuild]: @mjo I need your guidance with these.

File: net-misc/asterisk/asterisk-13.29.1.ebuild [hunk 19]
```diff
@@ -0,0 +1,322 @@
+# Copyright 1999-2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+inherit autotools linux-info systemd
+
+MY_P="${PN}-${PV/_/-}"
+
+DESCRIPTION="Asterisk: A Modular Open Source PBX System"
+HOMEPAGE="http://www.asterisk.org/"
```
Comment (juippis) [inline net-misc/asterisk/asterisk-13.29.1.ebuild]: `https://` when ever available, similarily to SRC_URI (if available).

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 2 above)
Comment (juippis) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: Ok, leave them like this for now. But when you introduce new one, let's follow the good guidelines.

File: acct-user/asterisk/asterisk-0.ebuild (hunk 18 above)
Comment (mjo) [reply acct-user/asterisk/asterisk-0.ebuild]: Wrong person? :)

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 1 above)
Comment (juippis) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: I meant in alphabetical order, that you seem to follow everywhere else in this ebuild...

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 3 above)
Comment (juippis) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: Hmm, yeah sure, sounds logical.

File: net-misc/asterisk/asterisk-13.29.1.ebuild (hunk 8 above)
Comment (jkroonza) [reply net-misc/asterisk/asterisk-13.29.1.ebuild]: Along with all of the ones higher up too ... wrapped up into a function.

File: acct-user/asterisk/asterisk-0.ebuild (hunk 18 above)
Comment (jkroonza) [reply acct-user/asterisk/asterisk-0.ebuild]: https://archives.gentoo.org/gentoo-dev/message/ce73d8d5170b1f6ec04e19ea3bdae267

Got some objections, but this one received no objections:

https://archives.gentoo.org/gentoo-dev/message/8b139b0c345da847f1f0bd95abb850fb

Comment (juippis) [reply acct-user/asterisk/asterisk-0.ebuild]: > Wrong person? :)

@mjo umm... seems so, sorry about the noise :)

Let's try this again: @orlitzky can we have your word on **defining these variables**, 
@jkroonza I don't doubt the GID+UID numbers are good, but I've seen a lot of ... preaching about defining ACCT_USER_HOME, however I haven't had time to study it closer by myself. So I'm consulting @orlitzky to explain it for both of us here :P

Comment (jkroonza) [reply acct-user/asterisk/asterisk-0.ebuild]: https://www.gentoo.org/glep/glep-0081.html#home-directory-ownership

So from that I understand:  Do not set ACCT_USER_HOME unless specifically required.

So I updated my test system to home folder /dev/null, and stopped and started asterisk with no problems.  So dropped from update.

Comment (orlitzky) [reply acct-user/asterisk/asterisk-0.ebuild]: Yep that's right. I've given up on trying to change the devmanual, so in the meantime I've [posted the guidelines on my website](http://michael.orlitzky.com/articles/gentoo_glep81_user_package_guidelines.xhtml) for easy copy/paste:

tl;dr use the defaults if possible
